tinybytes/src/lib.rs

```diff
@@ -291,6 +320,128 @@ impl Serialize for Bytes {
+    /// operation on the Waker, the associated function in the vtable of the underlying RawWaker
+    /// will be called.
+    ///
+    /// # Safety
+    ///
+    /// * The value pointed to by `data` must be 'static + Send + Sync
+    pub const unsafe fn from_raw(data: NonNull<()>, vtable: &'static RefCountedCellVTable) -> Self {
+        RefCountedCell {
+            raw: RawRefCountedCell { data, vtable },
+        }
+    }
+}
+
+impl Clone for RefCountedCell {
+    fn clone(&self) -> Self {
+        unsafe { (self.raw.vtable.clone)(self.raw.data.as_ptr().cast_const()) }
+    }
+}
+
+impl Drop for RefCountedCell {
+    fn drop(&mut self) {
+        unsafe { (self.raw.vtable.drop)(self.raw.data.as_ptr().cast_const()) }
+    }
+}
+
+struct RawRefCountedCell {
+    data: NonNull<()>,
+    vtable: &'static RefCountedCellVTable,
+}
+
+pub struct RefCountedCellVTable {
+    pub clone: unsafe fn(*const ()) -> RefCountedCell,
+    pub drop: unsafe fn(*const ()),
+}
+
+/// Creates a refcounted cell.
+///
+/// The data passed to this cell will only be dopped when the last
+/// clone of the cell is dropped.
+fn make_refcounted<T: Send + Sync + 'static>(data: T) -> RefCountedCell {
+    /// A custom Arc implementation that contains only the strong count
+    ///
+    /// This struct is not exposed to the outside of this functions and is
+    /// only interacted with through the `RefCountedCell` API.
+    struct CustomArc<T> {
+        rc: AtomicUsize,
+        #[allow(unused)]
+        data: T,
+    }
+
+    unsafe fn custom_arc_clone<T>(data: *const ()) -> RefCountedCell {
+        let custom_arc = data as *const CustomArc<T>;
+        let rc = unsafe {
+            std::ptr::addr_of!((*custom_arc).rc)
+                .as_ref()
+                .unwrap_unchecked()
+        };
+        rc.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+        RefCountedCell::from_raw(
+            NonNull::new_unchecked(data as *mut ()),
+            &RefCountedCellVTable {
+                clone: custom_arc_clone::<T>,
+                drop: custom_arc_drop::<T>,
+            },
+        )
+    }
+
+    unsafe fn custom_arc_drop<T>(data: *const ()) {
+        let custom_arc = data as *const CustomArc<T>;
+        let rc: &AtomicUsize = unsafe {
+            std::ptr::addr_of!((*custom_arc).rc)
+                .as_ref()
+                .unwrap_unchecked()
+        };
+        if rc.fetch_sub(1, std::sync::atomic::Ordering::Release) != 1 {
+            return;
+        }
+        {
+            let custom_arc = (custom_arc as *mut CustomArc<T>)
+                .as_mut()
+                .unwrap_unchecked();
+            std::ptr::drop_in_place(custom_arc);
+        }
+        // See standard library documentation for std::sync::Arc to see why this is needed.
+        // https://github.com/rust-lang/rust/blob/2a5da7acd4c3eae638aa1c46f3a537940e60a0e4/library/alloc/src/sync.rs#L2647-L2675
+        std::sync::atomic::fence(std::sync::atomic::Ordering::Acquire);
```

Comment:
Isn't the fence supposed to happen _before_ the drop_in_place?

Reply:
And then we can simplify this to `drop(Box::from_raw(data.as_ptr()));`

Reply:
>Isn't the fence supposed to happen before the drop_in_place

Yes, good catch 😅

Reply:
>And then we can simplify manual drop+dealloc

Actually no, this makes miri completely fail. I think it's not happy that we recreate a Box from a pointer it considers "shared", whereas it has no issues with doing this manually.
The Arc in std also creates it's allocation with Box::new, and frees with drop_in_place + dealloc

Reply:
Alright, then please note that miri is dumb and that's why it's done this way.